File: python/rail-fence-cipher/rail_fence_cipher.py
def fence_pattern(rails, message_size):
    pattern = [[] for _ in range(rails)]
    row = 0
    row_offset = 1
    for col in range(message_size):
        pattern[row].append(col)

        row += row_offset
        if not 0 <= row < len(pattern):
            row_offset *= -1
            row += row_offset * 2
    return pattern


def encode(message, rails):
    pattern = fence_pattern(rails, len(message))
    return ''.join([''.join([message[col] for col in line]) for line in pattern])


def decode(encoded_message, rails):
    pattern = fence_pattern(rails, len(encoded_message))
    result = [None] * len(encoded_message)
    index = 0
    for line in pattern:
        for col in line:
            result[col] = encoded_message[index]
            index += 1
    return ''.join(result)

File: haskell/rail-fence-cipher/src/RailFenceCipher.hs
module RailFenceCipher
  ( encode
  , decode
  ) where

encode :: Int -> String -> String
encode rails message =
  let pattern = fencePattern rails $ length message
   in concat [[message !! col | col <- line] | line <- pattern]

decode :: Int -> String -> String
decode rails encodedMessage =
  let pattern = fencePattern rails $ length encodedMessage
      (result, _) =
        foldl f (replicate (length encodedMessage) undefined, 0) pattern
   in result
  where
    f (result, index) line = foldl f1 (result, index) line
      where
        f1 (result, index) col =
          let (xs, _:ys) = splitAt col result
           in (xs ++ (encodedMessage !! index) : ys, index + 1)

fencePattern :: Int -> Int -> [[Int]]
fencePattern rails messageSize =
  let (pattern, _, _) =
        foldl f (replicate rails [], 0, 1) [0 .. messageSize - 1]
   in pattern
  where
    f (pattern, row, rowOffset) col
      | row < 0 || row >= length pattern =
        f (pattern, row - rowOffset * 2, -rowOffset) col
      | otherwise =
        ( [ (pattern !! r) ++
          if r == row
            then [col]
            else []
          | r <- [0 .. rails - 1]
          ]
        , row + rowOffset
        , rowOffset)
